core/src/code.rs

```diff
@@ -91,3 +91,62 @@ impl CheckedCodeWithHash {
         (self.0, self.1)
     }
 }
+
+/// Contains instrumented binary code of a program and initial memory size from memory import.
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq)]
+pub struct InstrumentedCode {
+    code: Vec<u8>,
+    static_pages: u32,
+    instruction_weights_version: u32,
+}
+
+impl InstrumentedCode {
+    /// Creates new instance with specified static pages and version of instruction weights.
+    pub fn new(code: Vec<u8>, static_pages: u32, instruction_weights_version: u32) -> Self {
+        Self {
+            code,
+            static_pages,
+            instruction_weights_version,
+        }
+    }
+
+    /// Returns reference to the instrumented binary code.
+    pub fn code(&self) -> &[u8] {
+        &self.code
+    }
+
+    /// Returns initial memory size from memory import.
+    pub fn static_pages(&self) -> u32 {
+        self.static_pages
+    }
+
+    /// Returns instruction weights version.
+    pub fn instruction_weights_version(&self) -> u32 {
+        self.instruction_weights_version
+    }
+}
+
+/// Contains instumented code for a program and the hash for it.
+pub struct InstrumentedCodeWithHash(InstrumentedCode, CodeId);
+
+impl InstrumentedCodeWithHash {
+    /// Creates new instance from the provided code.
+    pub fn new(code: InstrumentedCode, hash: CodeId) -> Self {
+        Self(code, hash)
+    }
```

Comment:
```suggestion
    pub fn new(code: InstrumentedCode) -> Self {
        let hash = CodeId::generate(code.checked_code().code());
        Self(code, hash)
    }
```

Reply:
and `From<InstrumentedCodeWithHash> for CheckedCodeWithHash` could be implemented. Frankly speaking I don't know if such conversion is useful at the moment:
```rust
impl From<InstrumentedCodeWithHash> for CheckedCodeWithHash {
    fn from(h: InstrumentedCodeWithHash) -> Self {
         let InstrumentedCode{ checked_code, .. } = h.0;
         Self(checked_code, h.1)
    }
}
```

Reply:
I will leave one structure `Code`. Already refactoring this.

Reply:
> I will leave one structure `Code`. Already refactoring this.

Do we really need Code* and Code*WithHash structures both? I don't see any reason to keep hash inside the structure, because hash is it's key

Reply:
Yep, we need. This new type guarantees that hash is calculated for the very same code. As a positive consequence - the hash is calculated only once.

Reply:
> Yep, we need. This new type guarantees that hash is calculated for the very same code. As a positive consequence - the hash is calculated only once.

Ah, ok. I thought we store them in storage also. Checked it again it we don't.